### ISSUE TITLE
 Inquiry Regarding InterCTC Model Experiments on AISHELL-1 Dataset

### ISSUE THREAD

**Issue by theshyPika** [espnet/espnet#5868]
Hi, 

I have conducted experiments based on the InterCTC model on the AISHELL-1 dataset, with auxiliary CTC losses calculated at layers [3, 6, 9, 12, 15]. I currently have the following three questions:

In the train.log for the best epoch, the ctc_cer for the [valid] validation set is higher than the cer in both dev and test. Is this because the validation set is not derived from dev and test?

In the results for stage 12, I obtained the inference results for each sub-layer [3, 6, 9, 12, 15], but it seems there are no cer statistics for each sub-layer. How can I obtain these, and do I need to implement this myself? How should I use the inference results to calculate it using which script?

Regarding RTF, I conducted the experiments on a cloud server using a GPU for inference, and the measured RTF for inference reached around 6.0 or even up to 15, which seems unusual because I've noticed that RTF in papers is generally less than 1. What could be the possible reason for this discrepancy? Although I am using a 3090 Ti, there shouldn't be such a significant difference, should there?

Thank you for your attention and assistance.

**Comment by sw005320** [comment]
> Hi,
> 
> I have conducted experiments based on the InterCTC model on the AISHELL-1 dataset, with auxiliary CTC losses calculated at layers [3, 6, 9, 12, 15]. I currently have the following three questions:
> 
> In the train.log for the best epoch, the ctc_cer for the [valid] validation set is higher than the cer in both dev and test. Is this because the validation set is not derived from dev and test?

Correct

> In the results for stage 12, I obtained the inference results for each sub-layer [3, 6, 9, 12, 15], but it seems there are no cer statistics for each sub-layer. How can I obtain these, and do I need to implement this myself? How should I use the inference results to calculate it using which script?

We do not prepare it for each sub-layer.
It is better to implement it by yourself.

> Regarding RTF, I conducted the experiments on a cloud server using a GPU for inference, and the measured RTF for inference reached around 6.0 or even up to 15, which seems unusual because I've noticed that RTF in papers is generally less than 1. What could be the possible reason for this discrepancy? Although I am using a 3090 Ti, there shouldn't be such a significant difference, should there?

How did you compute RTF?
We could achieve RTF < 1 (even for the CPU).

**Comment by theshypig** [comment]
Thanks for your reply.
1. In asr.h, It likes to compute cer by the tool named sclite ,so if I can manually use this tool to compute by myself.
2. [calculate_rtf.log](https://github.com/user-attachments/files/16597025/calculate_rtf.log)
Just by run.sh. I get this log file in logdir. The RTF result is wired.

**Comment by sw005320** [comment]
Thanks, but I still could not figure it out.
Could you give me the training/inference configs and inference log?

**Comment by theshypig** [comment]
As you requested, I have attached the log files for your review. I have also been reflecting on the performance issues we've been encountering and would like to share some of my speculations:

Initially, I suspected that the issue might be related to NLTK, as it seems to be associated with the cloud server environment I am using, which redownloads the NLTK data each time before inference. However, even after preparing the NLTK environment for inference testing, I still observed a high RTF (Real-Time Factor).

Regarding the BatchSize, I have noticed that it currently only supports a value of 1. I am wondering if this could be a contributing factor to the issue we are facing.

In my experiments, I utilized a 5-layer intermediate CTC and noticed that all five layers produced outputs during inference. I am considering whether this could be related to the observed performance.

In addition to the RTF concerns, I have encountered a new issue regarding the Transformer Language Model (LM) used for shallow fusion on the AISHELL dataset. I am seeking advice on better training configurations for the LM and recommendations for the lm_weight parameter. When I decoded using a shallow fusion of the LM and CTC, the decoding performance was not satisfactory, and in some cases, it was worse than my original results. For example:

Without LM: The results were 4.3 and 4.7.
With LM (default configuration, LM_weight=0.3, CTC_weight=1.0, Beam Search=10): The results were 4.4 and 4.7.
I suspect that the LM may not have been trained optimally, or the weight distribution is not appropriate. I would appreciate any recommendations for improved LM configurations that you might have.
[calculate_rtf (4).log](https://github.com/user-attachments/files/16660345/calculate_rtf.4.log)
[asr_inference.1 (2).log](https://github.com/user-attachments/files/16660346/asr_inference.1.2.log)
[train (1).log](https://github.com/user-attachments/files/16660375/train.1.log)
[run_interctc_train (1).log](https://github.com/user-attachments/files/16660378/run_interctc_train.1.log)
[config.txt](https://github.com/user-attachments/files/16660435/config.txt)

**Comment by sw005320** [comment]
I found that the GPU is not used in the inference log `--ngpu 0`
Also, please make sure to set the beam size 1 for CTC.

Also, an LM is not very strong in this task as the text size is very small.
We may get better performance with more LM tuning, but I'm not surprised by this result since intermediate-CTC can implicitly incorporate the LM effect.

**Comment by theshypig** [comment]
I apologize for the oversight; the results I mentioned were from an experiment conducted a month or two ago, which I failed to notice at the time.

However, even with GPU utilization, the RTF remains above 6. I've attached my inference logs for your reference.
[asr_inference.1 (5).log](https://github.com/user-attachments/files/16663710/asr_inference.1.5.log)
[calculate_rtf (5).log](https://github.com/user-attachments/files/16663711/calculate_rtf.5.log)

**Comment by sw005320** [comment]
It might be a bug (sorry, in this case), some files are extremely slow, or wrong usage when calculating RTF.
As you can see from the log, the utterance computation is finished in less than 0.5 seconds, while the audio length is around 4 seconds.
```
2024-05-03 12:16:33,577 (beam_search:483) INFO: best hypo: 广州市房地产中介协会分析

2024-05-03 12:16:33,813 (asr_inference:489) INFO: speech length: 432
2024-05-03 12:16:33,840 (beam_search:428) INFO: decoder input length: 107
2024-05-03 12:16:33,840 (beam_search:429) INFO: max output length: 107
2024-05-03 12:16:33,840 (beam_search:430) INFO: min output length: 0
2024-05-03 12:16:33,949 (beam_search:447) INFO: no hypothesis. Finish decoding.
2024-05-03 12:16:33,950 (beam_search:476) INFO:  -0.00 * 1.0 =  -0.00 for ctc
2024-05-03 12:16:33,950 (beam_search:479) INFO: total log probability: -0.00
2024-05-03 12:16:33,950 (beam_search:480) INFO: normalized log probability: -0.00
2024-05-03 12:16:33,950 (beam_search:481) INFO: total number of ended hypotheses: 1
2024-05-03 12:16:33,950 (beam_search:483) INFO: best hypo: 广州市房地产中介协会还表示
```
